src/Combinatorics/EnumerativeCombinatorics/multipartitions.jl

```diff
@@ -0,0 +1,166 @@
+  return length(MP.mp)
+end
+
+function Base.getindex(MP::Multipartition, i::Int)
+  return getindex(MP.mp,i)
+end
+
+
+"""
+    sum(P::Multipartition{T}) where T<:IntegerUnion
+
+If `P` is a multipartition of the integer n, this function returns n.
+"""
+function Base.sum(MP::Multipartition{T}) where T<:IntegerUnion
+  return sum(sum, MP.mp)
+end
+
+
+"""
+    multipartitions(n::T, r::IntegerUnion)  where T<:IntegerUnion
+
+A list of all `r`-component multipartitions of `n`, as elements of type Multipartitoon{T}.
+The algorithm is recursive and  based on [`partitions(::IntegerUnion)`](@ref).
+
+# Example
+```jldoctest
+julia> multipartitions(2,2)
+5-element Vector{Multipartition{Int64}}:
+ Partition{Int64}[[], [2]]
+ Partition{Int64}[[], [1, 1]]
+ Partition{Int64}[[1], [1]]
+ Partition{Int64}[[2], []]
+ Partition{Int64}[[1, 1], []]
+```
+"""
+function multipartitions(n::T, r::IntegerUnion) where T<:IntegerUnion
+  #Argument checking
+  @req n >= 0 "n >= 0 required"
+  @req r >= 1 "r >= 1 required"
+
+  #This will be the list of multipartitions
+  MP = Multipartition{T}[]
+
+  #We will go through all compositions of n into r parts, and for each such #composition, we collect all the partitions for each of the components of
+  #the composition.
+  #We create the compositions here in place for efficiency.
+
+  #recursively produces all Integer Vectors p of length r such that the sum of all the Elements equals n. Then calls recMultipartitions!
+  function recP!(p::Vector{T}, i::T, n::T) #where T<:IntegerUnion
+    if i==length(p) || n==0
+      p[Int(i)] = n
+      recMultipartitions!(fill(Partition(T[]),r), p, T(1))
+    else
+      for j=0:n
+        p[Int(i)] = T(j)
+        recP!(copy(p), T(i+1), T(n-j))
+      end
+    end
+  end
+
+  #recursively produces all multipartitions such that the i-th partition sums up to p[i]
+  function recMultipartitions!(mp::Vector{Partition{T}}, p::Vector{T}, i::T) #where T<:IntegerUnion
+    if i == length(p)
+      for q in partitions(p[Int(i)])
+        mp[Int(i)] = q
+        push!(MP, Multipartition{T}(copy(mp)))
+      end
+    else
+      for q in partitions(p[Int(i)])
+        mp[Int(i)] = q
+        recMultipartitions!(copy(mp), p, T(i+1))
+      end
+    end
+  end
+
+  recP!(zeros(T,r), T(1), n)
+  return MP
+end
+
+
+@doc raw"""
+    number_of_multipartitions(n::IntegerUnion, k::IntegerUnion)
+
+The number of multipartitions of ``n`` into ``k`` parts is equal to
+```math
+\sum_{a=1}^k {k \choose a} \sum_{λ} p(λ₁) p(λ₂) ⋯ p(λ_a) \;,
+```
+where the second sum is over all [compositions](@ref Compositions) ``λ`` of ``n`` into ``a`` parts. I found this formula in the Proof of Lemma 2.4 in [Cra06](@cite).
+"""
+function number_of_multipartitions(n::IntegerUnion, k::IntegerUnion)
+
+  # Special cases
+  n == 0 && return ZZ(k)
+
+  z = ZZ(0)
+  for a in 1:k
+    w = ZZ(0)
+    for lambda in compositions(n,a)
+      w += prod([number_of_partitions(lambda[i]) for i in 1:length(lambda)])
+    end
```

Comment:
```suggestion
    w = sum(lambda -> prod(number_of_partitions, lambda), compositions(n, a))
```
this saves some allocations